src/main/kotlin/org/rust/ide/inspections/RsNeedlessLifetimesInspection.kt

```diff
@@ -0,0 +1,240 @@
+    if (inputLifetimes.isEmpty()) return false
+
+    // no output lifetimes, check distinctness of input lifetimes
+    if (outputLifetimes.isEmpty()) {
+        // only unnamed and static, ok
+        if (inputLifetimes.all { it !is ReferenceLifetime.Named }) return false
+        // we have no output reference, so we only need all distinct lifetimes
+        return inputLifetimes.size == countUniqueLifetimes(inputLifetimes)
+    }
+
+    return when {
+        countUniqueLifetimes(outputLifetimes) > 1 -> false
+        inputLifetimes.size == 1 -> {
+            val input = inputLifetimes.first()
+            val output = outputLifetimes.first()
+            when {
+                input is ReferenceLifetime.Named && output is ReferenceLifetime.Named &&
+                    input.name == output.name -> true
+                input is ReferenceLifetime.Named && output == ReferenceLifetime.Unnamed -> true
+                else -> false  // already elided, different named lifetimes or something static going on
+            }
+        }
+        else -> false
+    }
+}
+
+private fun allowedLifetimesFrom(lifetimeParameters: List<RsLifetimeParameter>): Set<ReferenceLifetime> {
+    val allowedLifetimes = HashSet<ReferenceLifetime>()
+    lifetimeParameters
+        .filter { it.bounds.isEmpty() }
+        .mapNotNull { it.name }
+        .map { ReferenceLifetime.Named(it) }
+        .toCollection(allowedLifetimes)
+    allowedLifetimes.add(ReferenceLifetime.Unnamed)
+    allowedLifetimes.add(ReferenceLifetime.Static)
+    return allowedLifetimes
+}
+
+private fun lifetimesFromBounds(bounds: List<RsLifetime>): List<ReferenceLifetime> =
+    bounds.map { it.name }.filterIsInstance<ReferenceLifetime.Named>()
+
+private fun countUniqueLifetimes(lifetimes: List<ReferenceLifetime>): Int = HashSet(lifetimes).size
+
+/**
+ * Are any lifetimes mentioned in the `where` clause? If yes, we don't try to reason about elision.
+ */
+private fun hasWhereLifetimes(whereClause: RsWhereClause?): Boolean {
+    if (whereClause == null) return false
+    for (predicate in whereClause.wherePredList) {
+        if (predicate.lifetime != null) return true
+        // a predicate like F: Trait or F: for<'a> Trait<'a>
+
+        // check the type F, it may not contain LT refs
+        val ty = predicate.typeReference?.type
+        if (ty?.hasReEarlyBounds == true) return true
+
+        // if the bounds define new lifetimes, they are fine to occur
+        val boundLifetimeParams = predicate.forLifetimes?.lifetimeParameterList.orEmpty()
+        val allowedLifetimes = allowedLifetimesFrom(boundLifetimeParams)
+        // now walk the bounds
+        val collector = LifetimesCollector()
+        predicate.typeParamBounds?.polyboundList?.map { it.bound }?.forEach { it.accept(collector) }
+        if (collector.abort) return false
+        // and check that all lifetimes are allowed
+        if (collector.lifetimes.any { !allowedLifetimes.contains(it) }) return false
```

Comment:
I guess you can use `none` instead of `any` and remove this negation. I didn't even noticed it, it is really subtle :sweat_smile:

Reply:
Or you can use 
```kotlin
if (!allowedLifetimes.containsAll(collector.lifetimes)) return false
```
It is more concise